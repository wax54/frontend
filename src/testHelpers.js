import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import UserContext from './UserContext';

const validUserContext = {
    user: {
        username: "wax54",
        firstName: "Sam",
        lastName: "Crewe-Sullam",
        email: "wax654@gmail.c om",
        isAdmin: false,
        applications: [1, 2, 3]
    },
    isLoggedIn: true
}

function preRender(options={}) {
    const defaultOptions = {
        initialEntries: ['/'],
        context: {
            user: {},
            isLoggedIn: false
        }
    }
    const newOptions = {
        initialEntries: 
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        context: 
            { ...defaultOptions.context, ...(options.context || {}) }
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

function renderWithRouter(Component, options) {
    options = preRender(options);
    return render(
        <MemoryRouter initialEntries={options.initialEntries}>
            <Component />
        </MemoryRouter>
    );
}

async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <MemoryRouter initialEntries={options.initialEntries}>
            <Component />
        </MemoryRouter>
    ));
}


function renderWithUserContext(Component, options) {
    options = preRender(options);
    return render(
        <MemoryRouter initialEntries={options.initialEntries}>
            <UserContext.Provider value={options.context} >
                <Component />
            </UserContext.Provider>
        </MemoryRouter>
    );
}

async function renderWithUserContextAndWait(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <MemoryRouter initialEntries={options.initialEntries}>
            <UserContext.Provider value={options.context} >
                <Component />
            </UserContext.Provider>
        </MemoryRouter>
    ));
}



export { validUserContext, 
    renderWithRouter, 
    renderAndWaitFor, 
    renderWithUserContext, 
    renderWithUserContextAndWait }