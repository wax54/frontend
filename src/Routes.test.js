import { validUserContext, renderWithUserContext, renderWithUserContextAndWait } from './testHelpers';
import Routes from './Routes';

test("doesn't blow up", () => {
    renderWithUserContext(Routes);
    renderWithUserContext(Routes, {context: validUserContext});
});

test("matchesSnapshot", () => {
    const { asFragment } = renderWithUserContext(Routes);
    expect(asFragment()).toMatchSnapshot();
});

test("matchesSnapshot with valid user", () => {
    const { asFragment } = renderWithUserContext(Routes, { context: validUserContext });
    expect(asFragment()).toMatchSnapshot();
})

test("can navigate to different routes", () => {
    const { queryByText } = renderWithUserContext(
        Routes, {
        initialEntries: ['/login']
    });
    expect(queryByText('Login')).toBeInTheDocument();

});

test("can access protected Routes when logged in", () => {
    const { queryByText } = renderWithUserContext(
        Routes, { 
            context: validUserContext, 
            initialEntries: ['/companies'] 
        });
    expect(queryByText('Loading Companies...')).toBeInTheDocument();
});

test("can't access protected Routes when not logged in", () => {
    const { queryByText} = renderWithUserContext(
        Routes, 
        {initialEntries: ['/companies'] }
    );
    expect(queryByText('Loading Companies...')).not.toBeInTheDocument();
});

test("protected Route companies redirects to Home when not logged in", async () => {
    const { queryByText } = await renderWithUserContextAndWait(
        Routes,
        { initialEntries: ['/companies'] }
    );
    expect(queryByText('Please Login or Signup above!')).toBeInTheDocument();
});