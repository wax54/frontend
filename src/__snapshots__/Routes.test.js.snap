// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchesSnapshot 1`] = `
<DocumentFragment>
  <div
    class="Home"
  >
    <h1>
      Please Login or Signup above!
    </h1>
  </div>
</DocumentFragment>
`;

exports[`matchesSnapshot with valid user 1`] = `
<DocumentFragment>
  <div
    class="Home"
  >
    <h1>
      Hello and Welcome wax54!
    </h1>
  </div>
</DocumentFragment>
`;
